README.md
## Youloge.ueditor(定制版本：需要搭配使用)
=====

> 试了很多富文本编辑器,还是微信富文本编辑器+各种插件+样式中心可以满足需求：
> 所以开一个项目慢慢写：[另一个定制tinymce](https://github.com/youfeed/youloge.tinymce)
> 基于百度ueditor1.4.3.3 富文本编辑器(_ueditor.1.4.3.3)

## 更新日志

- 1.6.0 插件初始流程`正常插件 - 登录插件 - 图片上传签名 - 文档列表插件 - 选中文档 - 同步草稿 - 同步上传签名`
- 1.5.6 精简代码 重构在插件中初始化逻辑
- 1.5.5 增加`plugins`参数 允许静默初始化插件
- 1.5.2 继续抽离代码：在重构结构，像`tinymce 插件方向`靠齐
- 1.5.1 增强结构 清楚更多插件，改写部分插件 允许外部触发
- 1.5.0 改变目录解构`正式开发从这里开始` 打包`Releases 发布`
- 1.4.5 把`pageage.json` 升级了，支持`es6写法` 删除修改了一些插件
- 1.4.4 百度UEditor 1.4.4 版本

## 入门部署和体验 ##

```js
1. `git clone ` 仓库
2. `npm install` 安装依赖：`安装 grunt` 安装 `http-server`
3. 执行 `npm run build` 打包编译
4. 执行 `npm run dev` 开启本地预览
5. 重复 `npm run build` 页面刷新查看效果
```

## 参考项目
- [ueditor-plus](https://gitee.com/modstart-lib/ueditor-plus) 由`ModStart`开发
- [neditor](https://gitee.com/notadd/neditor) 2年前我最早接触的二开
- [jian27com/ueditor-plus-typecho](https://github.com/jian27com/ueditor-plus-typecho) UEditor-for-Typecho 二开
- [LinkPoly/UEditor-for-Typecho](https://github.comLinkPoly/UEditor-for-Typecho) UEditor-for-Typecho
:::


![image](https://github.com/youfeed/youloge.ueditor/assets/22223510/334f396c-f583-486e-b654-c8fdb71891d9)


这是第四次 再次整理富文本编辑器了：需要功能贴合，容易定制，样式自由并不简单

这里留个疑问：[请看ISSUE](https://github.com/youfeed/youloge.edotor/issues/1)

:::

changelog.md
#UEditor Change List


## 1.4.4

1. 只创建一个编辑器实例：默认ID `toolbar:#toolbar,editor:#editor`







##1.4.3
###bug修复&优化改进
1. 修复hasContents接口在非ie下只有空格时判断还为真的问题
2. 修复在粘贴word内容时，会误命中cm,pt这样的文本内容变成px的问题
3. 优化删除编辑器再创建编辑器时，编辑器的容器id发生变化的问题
4. 修复提交jsonp请求时，callback参数的xss漏洞
5. 新增jsp后台多种服务器配置下的路径定位
6. 修复ZeroClipboard的flash地址参数名称错误
7. 修复getActionUrl的bug
8. 整理配置参数，把遗漏在代码中的配置项整理到ueditor.config.js里
9. 修复图片拉伸工具和编辑拉伸长高器的样式冲突
10. 修复文字的unicode编码会被错误再次解析问题
11. 添加消息提示功能，冒泡提示信息
12. 优化上传功能提示，当后端配置项没正常加载，禁用上传功能
13. 修复单图上传按钮和jqueryValidate不兼容的问题
14. 简化了与jqueryValidate的结合操作，具体看_examples/jqueryValidateDemo.html
15. 修复在删除编辑器后，再次创建时丢失原有id的问题
16. 修复查找替换在一些块节点中会导致替换错误

##1.4.2
###重构前后端交互功能
1. 前端上传模块统一改用webuploader
2. 整体重构了文件上传的配置方式，改为统一在后端配置，前后端自动打通，[详细文档]()
3. 统一各上传模块的提交地址，各模块通过action参数区分类型，[详细文档]()
4. 提供serverparam命令，可在提交时追加任意参数，[详细文档]()
5. 统一了前端各上传模块的布局样式
6. 支持了在线附件预览和插入
7. 统一了后端返回数据格式，[详细文档]()
8. 各在线预览列表模块支持分组加载
9. 增加点击直接选择文件上传图片插件
10. 优化了粘贴图片的功能，上传时有loading和出错的提示
11. 添加jsonp的跨域请求支持

###bug修复&优化改进
1. 修复内容过多时，回到顶部，顶部内容被工具栏遮挡问题
2. 修复htmlparser解析标签的bug
3. 修复锚点是#开头时还添加http://的bug
4. 修复全屏后，退出全屏高度没有缩回来的问题
5. 文字选中后按delete删除，无法触发contentchange事件
6. 修复选择图片时点击8个控制点不拖动，contentchange会误触发的问题
7. 修复执行命令会触发多次contentchange事件
8. 修复输入文字，设置高度300，没有滚动条
9. 修复在不可编辑模式下，链接认可修改的问题
10. 修复全局css对于ueditor有时展现会出现按钮独占一行的问题
11. 修复在ie11下上传图片失败的问题
12. 修复chrome 34版本下报错问题
13. 修复ie8下插入多张远程图片之后，对话框假死问题
14. 修复uparse,在页面中多次调用某些模块无效果问题
15. 修复容器宽度100%时,改变窗口大小,宽度不自适应的问题
16. 将桥接ui和编辑器的业务代码放到了新建的adapter目录，方便维护
17. 修复拖拽改变图片大小功能和bootstrap不兼容的问题
18. 修复在表格水平边框上拖拽，不能改变大小的问题
19. 修复在表格标题上可以向左向下合并单元格的问题
20. 修复grunt打包出错的问题
21. 优化ie11下的兼容问题，主要修复了表格下的诸多问题
22. 优化插入图片，添加原文件名作为alt属性

###新功能添加
1. 添加了enableContextMenu配置开关，开关右键菜单，默认为true
2. 添加disablePInList配置，指定产出的列表中是否嵌套P标签，默认是false
3. 添加retainOnlyLabelPasted配置，指定粘贴时是否是只保留标签模式，默认是false
4. 优化了添加toolbar上各类ui的方式，方便二次开发
6. 优化有时产出数据会带有 bookmark标签的脏数据问题
7. 添加LocalStorage本地存储工具
8. 优化自动排版功能，选项保存到localStorage或UserData
9. 添加右键菜单的复制粘贴的支持，非ie不提供粘贴功能

##1.3.6

###bug修复
1.	script/style标签内容，在ie下的编辑状态显示内容问题
2.	修复预览窗口没有滚动条问题 [出自](http://www.ueditorbbs.com/forum.php?mod=viewthread&tid=28231&extra=page%3D1)
3.	修复在ie67下自动寻址的问题
4.	修复ff下图片拖拽大小不触发contentchange问题
5.	修复注释被删除的问题，包括在script中的[出自](http://www.ueditorbbs.com/forum.php?mod=viewthread&tid=12509&extra=page%3D1)
6.	修复在源码模式下不能使用setContent的问题，[出自](http://www.ueditorbbs.com/forum.php?mod=viewthread&tid=26910&extra=page%3D1)
7.	修复在给定宽度为100%时，ie11在拖动窗口大小时，编辑区域不随着变化的问题
8.	修复在ie67下destroy方法调用报错
9.	修复在插入代码模式下，ie9+以上的浏览器键盘操作(enter/tab等)错误或者失效的问题
10.	修复不规则源码在ie下解析成文本的问题
11.	修复p标签在ie下嵌套的问题
12. 修复目录大纲更新事件导致失焦点的问题
13. 针对word粘贴列表到编辑器中自动转换为list标签，由于有误命中情况，所以默认关闭该功能。提供**autoTransWordToList**配置项(默认为false)，开发者可酌情配置。
14. 添加禁止表格嵌套的开关，解决excel中粘贴到编辑器中会有冗余的嵌套表格问题，配置项**disabledTableInTable**(默认是true)。
15. 过滤掉excel的表格粘贴到ie中时，会有bitmap的冗余占位图片问题

###功能更新
1.	支持视频上传插入，多浏览器自适配播放器

#1.3.5

###新增功能
1.	asp后台的支持
2.	添加本地自动保存功能
3.	增加数据可视化展示功能
4.	编辑器实例上添加isFocus,blur方法
5.	新增在chrome下针对图片的拖拽宽高功能
6.	新增在高端浏览器下，qq截图粘贴上传，拖拽图片到编辑上传
7.	添加表格插入列标题功能
8.	添加设置表格可排序功能，支持表格在预览页排序
9.  添加生成目录功能

###已有功能优化
1.	ie8以上版本使用w3cRange
2.	使用grunt作为打包工具
3.	修复了过滤规则对于script/style的内容的转码
5.	自动寻址功能重构
6.	修复下拉菜单高度问题
7.	针对ie默认带有的autolink功能，添加开启禁用选项，创建时传入autolink:false就可禁用ie的autolink功能
8.	支持插入动态地图
9.	**API文档更新**
10.	图片上传路径可配置，增加前后端路径验证
11.	对uparse进行了拆分重构
12.	随下载包提供各种功能说明文档
13. 背景颜色功能重构，可以在预览页显示背景
14. 重写了查找替换插件，解决ff下window.find方法失效的问题


##1.2.6.1
1.	查找替换支持正则表达式
2.	修复了ie67下初始化宽高给定百分比
3.	增加类似word中的快捷菜单，默认关闭
4.	针对默认过滤回转换div为p标签，提供了配置开关allowDivTransToP,默认为true
5.	修复了在ie下删除分割线后光标定位的问题
6.	提供了手动加载语言文件，避免ie下有时会因语言文件加载失败导致编辑器加载失败,提示"not import language file"的错误
7.	优化了编辑器初始化时获得contentWindow可能不存在的情况
8.	优化了编辑器加载自定义样式的问题，默认initialStyle传入的css样式优先级最高，其次是指定的外部css文件
9.	工具栏支持指定位置折行，'|'表示分割符,'||'表示折行
10.	表格操作功能升级，优化了对表格的拖拉及双击操作，并且支持IE6+浏览器。
11.	修复编辑器在禁用状态下仍然可以拖动表格边框的bug。
12.	修复了分割线不能删除的问题
13.	修复了初始化内容过多时，编辑器不自动长高，要点击编辑器才会长高的问题
14.	优化了添加字符边框的展示效果，避免出现重叠的问题
15.	修复下拉菜单超出屏幕的bug
16.	修复table属性初始化时table布局错误的bug
17.	优化了选择工具栏上下拉菜单类型的操作命令时，选区会有闪动的问题
18.	优化了关于swfupload的一个xss漏洞
19.	优化了对于ie9,10的支持


##1.2.6
1.	优化了编辑器路径的设置，可以不用手动设置路径，自动识别相关路径，解决路径设置繁琐的问题
2.	重写了过滤粘贴机制，采用黑白名单，可以书写符合自己需求的过滤规则，可以完全定义标签的属性，甚至是style上的某个属性及其数值
3.	优化了拖拽机制，处理浮动图片拖拽不能跟指定的某行对齐
4.	数据同步改为失去焦点就执行，可以不再使用sync方法手动同步数据
5.	添加了字体边框
6.	优化了backspace/del键的操作
7.	重写了插入代码功能，插入代码编写支持tab和回车键
8.	表格支持排序和隔行显示
9.	改使用closure的压缩工具
10.	优化了undo/redo操作
11.	优化了ui界面


##1.2.5
###新增功能
1.	table整体重构
2.	table支持插入表头和标题
3.	table支持拷贝
4.	table支持任意调整宽高
5.	table支持任意前插后插行列
6.	table键盘操作仿word用户体验
7.	添加table平均分布行、列
8.	添加table单元格对齐方式
9.	添加table对齐方式
10.	添加选中部分表格,点击backspace或delete删除功能
11.	重写表格属性、单元格属性dialog
12.	粘贴支持纯文本,源码,纯标签3个模式选择
13.	添加计算字数的getContentLength接口
14.	添加计算字数事件wordcount
15.	图片上传支持参数动态绑定
16.	重写了list功能，支持一,一),(一),1),(1),——等新的列表标签
17.	调整了list中tab键的逻辑
18.	添加了可以限制列表的层级
19.	全屏快捷键 ALT+Z
20.	添加了uparseplugin.js展示页加载器

###优化修复
1.	优化了插入代码功能
2.	ie下默认禁用源码模式下的代码高亮
3.	截图功能支持非ie浏览器
4.	修正了非ie下中文输入时回退不准确的问题
5.	改进了键盘输入时做回退的操作

##1.2.4
###新增功能
1.	官网新增API文档
2. CSS按照UI结构进行了模块化拆分
3. 新增皮肤切换功能，并提供一套新皮肤（可通过配置项theme来设置）
4. 新增编辑器容器拖动缩放功能,配置项为:scaleEnabled、minFrameWidth、minFrameHeight
5. 新增音乐插件
6. 增加了源码模式下,全屏按钮可以使用
7. 添加了UE.getEditor工厂方法
8. 添加了针对jquery配合使用的demo
9. 添加了针对jqueryValidation配合使用的demo
10. 添加了初始化编辑器宽高配置，配置为项:initialFrameWidth、initialFrameHeight

###优化修复
1. 修复涂鸦路径在配置时，添加参数时请求报错
2. 修复涂鸦opera下缩放不能使用
3. 修复编辑器全屏功能失效问题
4. codemirror版本升级到最新版
5. 对opera/safari的支持进行了进一步的优化
6. 优化了部分demo页的代码
7. 修改原来的minFrameHeight为拖动时的最小高度

##1.2.3
###新增功能
1. 新增国际化支持
2. 新增涂鸦功能
3. 新增大小写功能
4. 新增getAllHtml方法，可以将整个页面的内容打出来,可以在ueditor.configplugin.js里通过配置allHtmlEnabled,来配置在提交时是否使用getAllHtml来得到内容
5. 新增插入模板的功能
6. 新增背景功能
7. 新增UE.instants全局对象，下边挂接了所有实例化的编辑器对象
8. Editor下新增ready方法，当编辑器ready后执行传入的fn,如果编辑器已经ready好了，就马上执行fn
9. 新增topOffset配置参数，用于设置AutoFloat时工具栏距离顶部的高度
10. 新增sourceEditorFirst配置参数，用于控制编辑器初始化时是否显示成源码模式，默认为否
11. 新增在表格内实例化编辑器的demo
12. 新增getDialog(dialogName)接口，可以获取dialog对象。

###优化修复
1. chrome下会出现alt+tab切换时，导致选区位置不对
2. focus方法添加参数可以指向到内容末尾
3. 完全支持opera浏览器
4. 修复了表格中实例化编辑器时工具栏浮动错位问题
5. 优化了后台处理文件代码,文件夹按照日期自动生成


##1.2.2
1.	编辑器不可编辑时，可以配置哪些功能可以使用，例如全屏
2.	table的边框为0时，采用虚线显示
3.	修复firefox下插入大量代码时，代码格式显示不正确的问题

ueditor.config.js
/**
 * ueditor完整配置项
 * 可以在这里配置整个编辑器的特性
 */
/**************************提示********************************
 * 所有被注释的配置项均为UEditor默认值。
 * 修改默认配置请首先确保已经完全明确该参数的真实用途。
 * 主要有两种修改方案，一种是取消此处注释，然后修改成对应参数；另一种是在实例化编辑器时传入对应参数。
 * 当升级编辑器时，可直接使用旧版配置文件替换新版配置文件,不用担心旧版配置文件中因缺少新功能所需的参数而导致脚本报错。
 **************************提示********************************/

(function () {

    /**
     * 编辑器资源文件根路径。它所表示的含义是：以编辑器实例化页面为当前路径，指向编辑器资源文件（即dialog等文件夹）的路径。
     * 鉴于很多同学在使用编辑器的时候出现的种种路径问题，此处强烈建议大家使用"相对于网站根目录的相对路径"进行配置。
     * "相对于网站根目录的相对路径"也就是以斜杠开头的形如"/myProject/ueditor/"这样的路径。
     * 如果站点中有多个不在同一层级的页面需要实例化编辑器，且引用了同一UEditor的时候，此处的URL可能不适用于每个页面的编辑器。
     * 因此，UEditor提供了针对不同页面的编辑器可单独配置的根路径，具体来说，在需要实例化编辑器的页面最顶部写上如下代码即可。当然，需要令此处的URL等于对应的配置。
     * window.UEDITOR_HOME_URL = "/xxxx/xxxx/";
     */
    var URL = window.UEDITOR_HOME_URL || getUEBasePath();

    /**
     * 配置项主体。注意，此处所有涉及到路径的配置别遗漏URL变量。
     */
    window.UEDITOR_CONFIG = {
        selector:['#toolbar','#editor'],
        //为编辑器实例添加一个路径，这个不能被注释
        UEDITOR_HOME_URL: URL
        // 服务器统一请求接口路径
        ,serverUrl: URL + "php/controller.php"
        // 定义需要下载的插件(/plugins/插件名/plugin.js)
        ,plugins:['article','link','simpleupload','insertfile']
        // 工具栏由`plugins 派生出来`
        ,toolbar:[]
        ,toolbars: [
            'undo', 'redo','|','removeformat', 'formatmatch', '|', 'fontsize','bold', 'italic', 'underline', 'fontborder', 'strikethrough','forecolor', 'backcolor',
            'superscript', 'subscript', 'autotypeset', 'blockquote', 'pasteplain', '|',  'insertorderedlist', 'insertunorderedlist', 'selectall', 'cleardoc', '|',
            'rowspacingtop', 'rowspacingbottom', 'lineheight', '|',
            'customstyle', 'paragraph', 'fontfamily',  '|',
            'directionalityltr', 'directionalityrtl', 'indent', '|',// 'link', 'simpleupload',
            'justifyleft', 'justifycenter', 'justifyright', 'justifyjustify', '|', 'touppercase', 'tolowercase', '|',
            'unlink', '|', 'imagenone', 'imageleft', 'imageright', 'imagecenter', '|',
            'insertimage', 'emotion', 'scrawl', 'insertvideo', 'music', 'attachment', 'map', 'gmap', 'insertframe', 'insertcode', 'webapp', 'pagebreak', 'template', 'background', '|',
            'horizontal', 'date', 'time', 'spechars', 'wordimage', '|',
            'inserttable',  '|',
            'preview', 'searchreplace', '|', 'source',
        ]
        // 参考 `tinymce`进行配置 插件触发方式不一定非要从toolbar点击 允许分别加载 `toolbar` 按需展示
        
        // 
        //当鼠标放在工具栏上时显示的tooltip提示,留空支持自动多语言配置，否则以配置值为准
        
        //语言配置项,默认是zh-cn。有需要的话也可以使用如下这样的方式来自动多语言切换，当然，前提条件是lang文件夹下存在对应的语言文件：
        //lang值也可以通过自动获取 (navigator.language||navigator.browserLanguage ||navigator.userLanguage).toLowerCase()
        //,lang:"zh-cn"
        //,langPath:URL +"lang/"

        //主题配置项,默认是default。有需要的话也可以使用如下这样的方式来自动多主题切换，当然，前提条件是themes文件夹下存在对应的主题文件：
        //现有如下皮肤:default
        //,theme:'default'
        //,themePath:URL +"themes/"

        //,zIndex : 900     //编辑器层级的基数,默认是900

        //针对getAllHtml方法，会在对应的head标签中增加该编码设置。
        //,charset:"utf-8"

        //若实例化编辑器的页面手动修改的domain，此处需要设置为true
        //,customDomain:false

        //常用配置项目
        //,isShow : true    //默认显示编辑器

        //,textarea:'editorValue' // 提交表单时，服务器获取编辑器提交内容的所用的参数，多实例时可以给容器name属性，会将name给定的值最为每个实例的键值，不用每次实例化的时候都设置这个值

        //,initialContent:'欢迎使用ueditor!'    //初始化编辑器的内容,也可以通过textarea/script给值，看官网例子

        //,autoClearinitialContent:true //是否自动清除编辑器初始内容，注意：如果focus属性设置为true,这个也为真，那么编辑器一上来就会触发导致初始化的内容看不到了

        //,focus:false //初始化时，是否让编辑器获得焦点true或false

        //如果自定义，最好给p标签如下的行高，要不输入中文时，会有跳动感
        //,initialStyle:'p{line-height:1em}'//编辑器层级的基数,可以用来改变字体等

        //,iframeCssUrl: URL + '/themes/iframe.css' //给编辑区域的iframe引入一个css文件

        //indentValue
        //首行缩进距离,默认是2em
        //,indentValue:'2em'

        ,initialFrameWidth:768  //初始化编辑器宽度,默认1000
        ,initialFrameHeight:801  //初始化编辑器高度,默认320

        //,readonly : false //编辑器初始化结束后,编辑区域是否是只读的，默认是false

        //,autoClearEmptyNode : true //getContent时，是否删除空的inlineElement节点（包括嵌套的情况）

        //启用自动保存
        //,enableAutoSave: true
        //自动保存间隔时间， 单位ms
        //,saveInterval: 500


        //,imagePopup:true      //图片操作的浮层开关，默认打开

        //,autoSyncData:true //自动同步编辑器要提交的数据
        //,emotionLocalization:false //是否开启表情本地化，默认关闭。若要开启请确保emotion文件夹下包含官网提供的images表情文件夹

        //粘贴只保留标签，去除标签所有属性
        //,retainOnlyLabelPasted: false

        //,pasteplain:false  //是否默认为纯文本粘贴。false为不使用纯文本粘贴，true为使用纯文本粘贴
        //纯文本粘贴模式下的过滤规则
        //'filterTxtRules' : function(){
        //    function transP(node){
        //        node.tagName = 'p';
        //        node.setStyle();
        //    }
        //    return {
        //        //直接删除及其字节点内容
        //        '-' : 'script style object iframe embed input select',
        //        'p': {$:{}},
        //        'br':{$:{}},
        //        'div':{'$':{}},
        //        'li':{'$':{}},
        //        'caption':transP,
        //        'th':transP,
        //        'tr':transP,
        //        'h1':transP,'h2':transP,'h3':transP,'h4':transP,'h5':transP,'h6':transP,
        //        'td':function(node){
        //            //没有内容的td直接删掉
        //            var txt = !!node.innerText();
        //            if(txt){
        //                node.parentNode.insertAfter(UE.uNode.createText(' &nbsp; &nbsp;'),node);
        //            }
        //            node.parentNode.removeChild(node,node.innerText())
        //        }
        //    }
        //}()

        //,allHtmlEnabled:false //提交到后台的数据是否包含整个html字符串

        //insertorderedlist
        //有序列表的下拉配置,值留空时支持多语言自动识别，若配置值，则以此值为准
        //,'insertorderedlist':{
        //      //自定的样式
        //        'num':'1,2,3...',
        //        'num1':'1),2),3)...',
        //        'num2':'(1),(2),(3)...',
        //        'cn':'一,二,三....',
        //        'cn1':'一),二),三)....',
        //        'cn2':'(一),(二),(三)....',
        //     //系统自带
        //     'decimal' : '' ,         //'1,2,3...'
        //     'lower-alpha' : '' ,    // 'a,b,c...'
        //     'lower-roman' : '' ,    //'i,ii,iii...'
        //     'upper-alpha' : '' , lang   //'A,B,C'
        //     'upper-roman' : ''      //'I,II,III...'
        //}

        //insertunorderedlist
        //无序列表的下拉配置，值留空时支持多语言自动识别，若配置值，则以此值为准
        //,insertunorderedlist : { //自定的样式
        //    'dash' :'— 破折号', //-破折号
        //    'dot':' 。 小圆圈', //系统自带
        //    'circle' : '',  // '○ 小圆圈'
        //    'disc' : '',    // '● 小圆点'
        //    'square' : ''   //'■ 小方块'
        //}
        //,listDefaultPaddingLeft : '30'//默认的左边缩进的基数倍
        //,listiconpath : 'http://bs.baidu.com/listicon/'//自定义标号的路径
        //,maxListLevel : 3 //限制可以tab的级数, 设置-1为不限制

        //,autoTransWordToList:false  //禁止word中粘贴进来的列表自动变成列表标签

        //fontfamily
        //字体设置 label留空支持多语言自动切换，若配置，则以配置值为准
        //,'fontfamily':[
        //    { label:'',name:'songti',val:'宋体,SimSun'},
        //    { label:'',name:'kaiti',val:'楷体,楷体_GB2312, SimKai'},
        //    { label:'',name:'yahei',val:'微软雅黑,Microsoft YaHei'},
        //    { label:'',name:'heiti',val:'黑体, SimHei'},
        //    { label:'',name:'lishu',val:'隶书, SimLi'},
        //    { label:'',name:'andaleMono',val:'andale mono'},
        //    { label:'',name:'arial',val:'arial, helvetica,sans-serif'},
        //    { label:'',name:'arialBlack',val:'arial black,avant garde'},
        //    { label:'',name:'comicSansMs',val:'comic sans ms'},
        //    { label:'',name:'impact',val:'impact,chicago'},
        //    { label:'',name:'timesNewRoman',val:'times new roman'}
        //]

        //fontsize
        //字号
        //,'fontsize':[10, 11, 12, 14, 16, 18, 20, 24, 36]

        //paragraph
        //段落格式 值留空时支持多语言自动识别，若配置，则以配置值为准
        //,'paragraph':{'p':'', 'h1':'', 'h2':'', 'h3':'', 'h4':'', 'h5':'', 'h6':''}

        //rowspacingtop
        //段间距 值和显示的名字相同
        //,'rowspacingtop':['5', '10', '15', '20', '25']

        //rowspacingBottom
        //段间距 值和显示的名字相同
        //,'rowspacingbottom':['5', '10', '15', '20', '25']

        //lineheight
        //行内间距 值和显示的名字相同
        //,'lineheight':['1', '1.5','1.75','2', '3', '4', '5']

        //customstyle
        //自定义样式，不支持国际化，此处配置值即可最后显示值
        //block的元素是依据设置段落的逻辑设置的，inline的元素依据BIU的逻辑设置
        //尽量使用一些常用的标签
        //参数说明
        //tag 使用的标签名字
        //label 显示的名字也是用来标识不同类型的标识符，注意这个值每个要不同，
        //style 添加的样式
        //每一个对象就是一个自定义的样式
        //,'customstyle':[
        //    {tag:'h1', name:'tc', label:'', style:'border-bottom:#ccc 2px solid;padding:0 4px 0 0;text-align:center;margin:0 0 20px 0;'},
        //    {tag:'h1', name:'tl',label:'', style:'border-bottom:#ccc 2px solid;padding:0 4px 0 0;margin:0 0 10px 0;'},
        //    {tag:'span',name:'im', label:'', style:'font-style:italic;font-weight:bold'},
        //    {tag:'span',name:'hi', label:'', style:'font-style:italic;font-weight:bold;color:rgb(51, 153, 204)'}
        //]

        //打开右键菜单功能
        //,enableContextMenu: true
        //右键菜单的内容，可以参考plugins/contextmenu.js里边的默认菜单的例子，label留空支持国际化，否则以此配置为准
        //,contextMenu:[
        //    {
        //        label:'',       //显示的名称
        //        cmdName:'selectall',//执行的command命令，当点击这个右键菜单时
        //        //exec可选，有了exec就会在点击时执行这个function，优先级高于cmdName
        //        exec:function () {
        //            //this是当前编辑器的实例
        //            //this.ui._dialogs['inserttableDialog'].open();
        //        }
        //    }
        //]

        //快捷菜单
        //,shortcutMenu:["fontfamily", "fontsize", "bold", "italic", "underline", "forecolor", "backcolor", "insertorderedlist", "insertunorderedlist"]

        //wordCount
        //,wordCount:true          //是否开启字数统计
        //,maximumWords:10000       //允许的最大字符数
        //字数统计提示，{#count}代表当前字数，{#leave}代表还可以输入多少字符数,留空支持多语言自动切换，否则按此配置显示
        //,wordCountMsg:''   //当前已输入 {#count} 个字符，您还可以输入{#leave} 个字符
        //超出字数限制提示  留空支持多语言自动切换，否则按此配置显示
        //,wordOverFlowMsg:''    //<span style="color:red;">你输入的字符个数已经超出最大允许值，服务器可能会拒绝保存！</span>

        //tab
        //点击tab键时移动的距离,tabSize倍数，tabNode什么字符做为单位
        //,tabSize:4
        //,tabNode:'&nbsp;'

        //removeFormat
        //清除格式时可以删除的标签和属性
        //removeForamtTags标签
        //,removeFormatTags:'b,big,code,del,dfn,em,font,i,ins,kbd,q,samp,small,span,strike,strong,sub,sup,tt,u,var'
        //removeFormatAttributes属性
        //,removeFormatAttributes:'class,style,lang,width,height,align,hspace,valign'

        //undo
        //可以最多回退的次数,默认20
        //,maxUndoCount:20
        //当输入的字符数超过该值时，保存一次现场
        //,maxInputCount:1

        //autoHeightEnabled
        // 是否自动长高,默认true
        //,autoHeightEnabled:true

        //scaleEnabled
        //是否可以拉伸长高,默认true(当开启时，自动长高失效)
        //,scaleEnabled:false
        //,minFrameWidth:800    //编辑器拖动时最小宽度,默认800
        //,minFrameHeight:220  //编辑器拖动时最小高度,默认220

        //autoFloatEnabled
        //是否保持toolbar的位置不动,默认true
        //,autoFloatEnabled:true
        //浮动时工具栏距离浏览器顶部的高度，用于某些具有固定头部的页面
        //,topOffset:30
        //编辑器底部距离工具栏高度(如果参数大于等于编辑器高度，则设置无效)
        //,toolbarTopOffset:400

        //设置远程图片是否抓取到本地保存
        //,catchRemoteImageEnable: true //设置是否抓取远程图片

        //pageBreakTag
        //分页标识符,默认是_ueditor_page_break_tag_
        //,pageBreakTag:'_ueditor_page_break_tag_'

        //autotypeset
        //自动排版参数
        //,autotypeset: {
        //    mergeEmptyline: true,           //合并空行
        //    removeClass: true,              //去掉冗余的class
        //    removeEmptyline: false,         //去掉空行
        //    textAlign:"left",               //段落的排版方式，可以是 left,right,center,justify 去掉这个属性表示不执行排版
        //    imageBlockLine: 'center',       //图片的浮动方式，独占一行剧中,左右浮动，默认: center,left,right,none 去掉这个属性表示不执行排版
        //    pasteFilter: false,             //根据规则过滤没事粘贴进来的内容
        //    clearFontSize: false,           //去掉所有的内嵌字号，使用编辑器默认的字号
        //    clearFontFamily: false,         //去掉所有的内嵌字体，使用编辑器默认的字体
        //    removeEmptyNode: false,         // 去掉空节点
        //    //可以去掉的标签
        //    removeTagNames: {标签名字:1},
        //    indent: false,                  // 行首缩进
        //    indentValue : '2em',            //行首缩进的大小
        //    bdc2sb: false,
        //    tobdc: false
        //}

        //tableDragable
        //表格是否可以拖拽
        //,tableDragable: true



        //sourceEditor
        //源码的查看方式,codemirror 是代码高亮，textarea是文本框,默认是codemirror
        //注意默认codemirror只能在ie8+和非ie中使用
        //,sourceEditor:"codemirror"
        //如果sourceEditor是codemirror，还用配置一下两个参数
        //codeMirrorJsUrl js加载的路径，默认是 URL + "third-party/codemirror/codemirror.js"
        //,codeMirrorJsUrl:URL + "third-party/codemirror/codemirror.js"
        //codeMirrorCssUrl css加载的路径，默认是 URL + "third-party/codemirror/codemirror.css"
        //,codeMirrorCssUrl:URL + "third-party/codemirror/codemirror.css"
        //编辑器初始化完成后是否进入源码模式，默认为否。
        //,sourceEditorFirst:false

        //allowLinkProtocol 允许的链接地址，有这些前缀的链接地址不会自动添加http
        //, allowLinkProtocols: ['http:', 'https:', '#', '/', 'ftp:', 'mailto:', 'tel:', 'git:', 'svn:']

        //webAppKey 百度应用的APIkey，每个站长必须首先去百度官网注册一个key后方能正常使用app功能，注册介绍，http://app.baidu.com/static/cms/getapikey.html
        //, webAppKey: ""

        //默认过滤规则相关配置项目
        //,disabledTableInTable:true  //禁止表格嵌套
        //,rgb2Hex:true               //默认产出的数据中的color自动从rgb格式变成16进制格式

    };

    function getUEBasePath(docUrl, confUrl) {

        return getBasePath(docUrl || self.document.URL || self.location.href, confUrl || getConfigFilePath());

    }

    function getConfigFilePath() {

        var configPath = document.getElementsByTagName('script');

        return configPath[ configPath.length - 1 ].src;

    }

    function getBasePath(docUrl, confUrl) {

        var basePath = confUrl;


        if (/^(\/|\\\\)/.test(confUrl)) {

            basePath = /^.+?\w(\/|\\\\)/.exec(docUrl)[0] + confUrl.replace(/^(\/|\\\\)/, '');

        } else if (!/^[a-z]+:/i.test(confUrl)) {

            docUrl = docUrl.split("#")[0].split("?")[0].replace(/[^\\\/]+$/, '');

            basePath = docUrl + "" + confUrl;

        }

        return optimizationPath(basePath);

    }

    function optimizationPath(path) {

        var protocol = /^[a-z]+:\/\//.exec(path)[ 0 ],
            tmp = null,
            res = [];

        path = path.replace(protocol, "").split("?")[0].split("#")[0];

        path = path.replace(/\\/g, '/').split(/\//);

        path[ path.length - 1 ] = "";

        while (path.length) {

            if (( tmp = path.shift() ) === "..") {
                res.pop();
            } else if (tmp !== ".") {
                res.push(tmp);
            }

        }

        return protocol + res.join("/");

    }

    window.UE = {
        getUEBasePath: getUEBasePath
    };

})();
